daemon/service-db.cc

```diff
@@ -56,149 +74,277 @@ MLServiceDB::~MLServiceDB ()
 void
 MLServiceDB::connectDB ()
 {
-  char *err = nullptr;
-  leveldb_options_t *db_options;
+  int rc;
+  char *sql;
+  char *errmsg = nullptr;
 
-  if (db_obj)
+  if (_db != nullptr)
     return;
 
-  db_options = leveldb_options_create ();
-  leveldb_options_set_create_if_missing (db_options, 1);
+  rc = sqlite3_open (_path.c_str (), &_db);
+  if (rc != SQLITE_OK) {
+    g_warning ("Failed to open database: %s (%d)", sqlite3_errmsg (_db), rc);
+    goto error;
+  }
+
+  /* Create table and handle database version. */
+  if (!_initialized) {
+    /**
+     * @todo handle database version
+     * - check old level db and insert pipeline description into sqlite.
+     * - check database info (TBL_DB_INFO), fetch data and update table schema when version is updated, ...
+     * - need transaction
+     */
 
-  db_obj = leveldb_open (db_options, path.c_str (), &err);
-  leveldb_options_destroy (db_options);
-  if (err != nullptr) {
-    g_warning ("Error! Failed to open database located at '%s': leveldb_open() has returned an error: %s",
-        path.c_str (), err);
-    leveldb_free (err);
+    /* Create tables. */
+    sql = g_strdup_printf ("CREATE TABLE IF NOT EXISTS %s;", g_mlsvc_table_schema[TBL_DB_INFO]);
+    rc = sqlite3_exec (_db, sql, nullptr, nullptr, &errmsg);
+    g_free (sql);
+    if (rc != SQLITE_OK) {
+      g_warning ("Failed to create table for database info: %s (%d)", errmsg, rc);
+      sqlite3_clear_errmsg (errmsg);
+      goto error;
+    }
+
+    sql = g_strdup_printf ("CREATE TABLE IF NOT EXISTS %s;", g_mlsvc_table_schema[TBL_PIPELINE_DESCRIPTION]);
+    rc = sqlite3_exec (_db, sql, nullptr, nullptr, &errmsg);
+    g_free (sql);
+    if (rc != SQLITE_OK) {
+      g_warning ("Failed to create table for pipeline description: %s (%d)", errmsg, rc);
+      sqlite3_clear_errmsg (errmsg);
+      goto error;
+    }
+
+    sql = g_strdup_printf ("CREATE TABLE IF NOT EXISTS %s;", g_mlsvc_table_schema[TBL_MODEL_INFO]);
+    rc = sqlite3_exec (_db, sql, nullptr, nullptr, &errmsg);
+    g_free (sql);
+    if (rc != SQLITE_OK) {
+      g_warning ("Failed to create table for model info: %s (%d)", errmsg, rc);
+      sqlite3_clear_errmsg (errmsg);
```

Comment:
I can't find sqlite3_clear_errmsg on google.
I'd like want to know how this differs from sqlite3_free().

Reply:
It's defined in the file as a macro.

Reply:
Thank you so much for the explanation. :)